demo-circuit/src/naive_threshold_sig_w_key_rotation/tests/verify.rs

```diff
@@ -426,75 +509,153 @@ fn malicious_master_key_rotation() {
+    debug_naive_threshold_circuit(&circuit, false, None);
+
+    // now the attacker can try to mount the replay attack, re-updating the old signing key to the
+    // previous signing key
+    let to_be_restored_signing_key = validator_key_updates.signing_keys[0];
+    let prev_withdrawal_certificate = withdrawal_certificate;
+    withdrawal_certificate = create_withdrawal_certificate();
+    validator_key_updates.signing_keys[0] = validator_key_updates.updated_signing_keys[0];
+    validator_key_updates.master_keys[0] = validator_key_updates.updated_master_keys[0];
+    validator_key_updates.updated_signing_keys[0] = to_be_restored_signing_key;
+    validator_key_updates.updated_signing_keys_sk_signatures[0] = old_update_signature_with_sk;
+    validator_key_updates.updated_signing_keys_mk_signatures[0] = old_update_signature_with_mk;
+
+    withdrawal_certificate.custom_fields[0] = validator_key_updates.get_upd_validators_keys_root()
+        .unwrap();
+
+    let msg_to_sign = cert_to_msg(&withdrawal_certificate);
+    let wcert_signatures = create_signatures(
+        MAX_PKS,
+        &signing_keys_sks,
+        &validator_key_updates.signing_keys,
+        msg_to_sign,
+    );
 
-    // Now try without previous withdrawal certificate
     let circuit_res = NaiveThresholdSignatureWKeyRotation::new(
         validator_key_updates.clone(),
         wcert_signatures,
         withdrawal_certificate.clone(),
-        None,
+        Some(prev_withdrawal_certificate.clone()),
         THRESHOLD as u64,
         genesis_validator_keys_tree_root,
     );
     assert!(circuit_res.is_ok());
     let circuit = circuit_res.unwrap();
 
-    debug_naive_threshold_circuit(&circuit, true, Some("check key changes/check updated master key should be signed old master key 0/conditional verify signature/conditional_equals"));
-}*/
+    debug_naive_threshold_circuit(&circuit, true, Some("check key changes/check updated signing key should be signed old signing key 0/conditional verify signature/conditional_equals"));
+println!("ok");
+    // test that if epoch_id is changed, then the attack works
+    withdrawal_certificate.epoch_id = old_epoch_id;
+    let msg_to_sign = cert_to_msg(&withdrawal_certificate);
+    let wcert_signatures = create_signatures(
+        MAX_PKS,
+        &signing_keys_sks,
+        &validator_key_updates.signing_keys,
+        msg_to_sign,
+    );
+
+    let circuit_res = NaiveThresholdSignatureWKeyRotation::new(
+        validator_key_updates,
+        wcert_signatures,
+        withdrawal_certificate,
+        Some(prev_withdrawal_certificate),
+        THRESHOLD as u64,
+        genesis_validator_keys_tree_root,
+    );
+    assert!(circuit_res.is_ok());
+    let circuit = circuit_res.unwrap();
 
+    debug_naive_threshold_circuit(&circuit, false, None);
+}
 
 #[serial]
 #[test]
-fn multiple_custom_fields() {
-    const THRESHOLD: usize = 4;
-    const CUSTOM_FIELDS: usize = 5;
-
+fn test_key_update_from_other_key_update() {
```

Comment:
Can we add some comment about the motivation of this test ?

Reply:
Ok